src/terrain/clients/iplant_groups.clj

```diff
@@ -266,20 +271,49 @@
     (verify-group-exists client user group)
     (c/list-group-members client user group)))
 
+(defn- format-privilege-updates [subject-ids privileges]
+  {:updates (vec (for [subject-id subject-ids] {:subject_id subject-id :privileges privileges}))})
+
+(defn- grant-optout-privileges [client user group members]
+  (c/update-group-privileges client user group (format-privilege-updates members ["optout"]) {:replace false}))
+
+(defn- revoke-optout-privileges [client user group members]
+  (c/revoke-group-privileges client user group (format-privilege-updates members ["optout"])))
+
 (defn add-team-members [user name members]
   (let [client (get-client)
         folder (get-team-folder-name client)
         group  (full-group-name name folder)]
     (verify-group-exists client user group)
+    (grant-optout-privileges client user group members)
     (c/add-group-members client user group members)))
 
 (defn remove-team-members [user name members]
   (let [client (get-client)
         folder (get-team-folder-name client)
         group  (full-group-name name folder)]
     (verify-group-exists client user group)
+    (revoke-optout-privileges client user group members)
     (c/remove-group-members client user group members)))
 
+(defn join-team [user name]
+  (let [client (get-client)
+        folder (get-team-folder-name client)
+        group  (full-group-name name folder)]
+    (verify-group-exists client user group)
+    (let [response (c/add-group-members client user group [user])]
+      (grant-optout-privileges client (config/grouper-user) group [user])
+      response)))
+
+(defn leave-team [user name]
+  (let [client (get-client)
+        folder (get-team-folder-name client)
+        group  (full-group-name name folder)]
+    (verify-group-exists client user group)
+    (let [response (c/remove-group-members client user group [user])]
+      (revoke-optout-privileges client (config/grouper-user) group [user])
+      response)))
```

Comment:
The update order here serves a similar purpose to the update order for `join-team` it's not as important here, however, because a permission is being removed.